Add support for attaching interfaces to running instances

Elethiomel opened the pull request:
In the Openstack WebUI, it's possible to attach an interface to an instances. This PR adds a new resource "openstack_compute_attach_interface_v2" that adds support for that in terraform. 

I've tested this internally and it's working well for me.

TODO: Add support for FixIPs, add unit tests and import support.

jtopjian commented:
@Elethiomel Nice - thanks for submitting this :)

Quick question: did you use `govendor` to add the `vendor` files? I don't see a `vendor.json` file in the commits. If you didn't use `govendor`, do you mind re-doing that one commit? The steps will be something like:

```shell
$ govendor fetch github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/...
$ git add vendor
$ git commit -m "vendor: Updating Gophercloud for attachinterfaces
```

Let me know if you have any questions. :)

Elethiomel commented:
@jtopjian I didn't use govendor, sorry. I've never used it before.

I did what you suggested and it did indeed update the vendor/vendor.json, however it pulled in a new directory too: vendor/github.com/gophercloud/gophercloud/testhelper/

```
fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ govendor fetch github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/...                                                                                                                                 
fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        deleted:    vendor/github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/testing/requests_test.go
        modified:   vendor/vendor.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        vendor/github.com/gophercloud/gophercloud/testhelper/

no changes added to commit (use "git add" and/or "git commit -a")

```
What do you recommend  I do in this case?

fatmcgav commented:
@Elethiomel Is it possible that that directory existed prior to you running `govendor`? As from memory there shouldn't be any untracked files as a result of a `govendor` operation...

jtopjian commented:
@Elethiomel 

> I didn't use govendor, sorry. I've never used it before.

Absolutely no problem at all. This is something we need to outline in our README.

As @fatmcgav mentioned, maybe this directory existed before? The best way of resolving this might be to rewind your git repository and try again:

```shell
$ git reset HEAD^^   # go back two commits
$ git status         # make sure untracked changes look ok
$ rm -rf vendor
$ govendor fetch ...
$ git add vendor
$ git commit -m ...
$ git add .
$ git commit -m ...
```

Let me know if that helps.

Elethiomel commented:
I'm afraid it's definitely govendor adding the files. Here's my repo cleanly checked out. There's no testhelper directory

```
fowler@snoopy:~/Code/src/github.com/terraform-providers$  git clone git@github.com:Elethiomel/terraform-provider-openstack.git
Cloning into 'terraform-provider-openstack'...
remote: Counting objects: 7987, done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 7987 (delta 0), reused 2 (delta 0), pack-reused 7979
Receiving objects: 100% (7987/7987), 5.71 MiB | 8.73 MiB/s, done.
Resolving deltas: 100% (4070/4070), done.

fowler@snoopy:~/Code/src/github.com/terraform-providers$ cd terraform-provider-openstack/

fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ ls vendor/github.com/gophercloud/gophercloud/testhelper/
ls: cannot access 'vendor/github.com/gophercloud/gophercloud/testhelper/': No such file or directory

fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

```

So everything is fine. There's no testhelper directory.
Let's run govendor.

```
fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ govendor fetch github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/...                                               

fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ ls vendor/github.com/gophercloud/gophercloud/testhelper/
client  convenience.go  doc.go  http_responses.go

fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        deleted:    vendor/github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/testing/requests_test.go
        modified:   vendor/vendor.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        vendor/github.com/gophercloud/gophercloud/testhelper/

no changes added to commit (use "git add" and/or "git commit -a")


```
There's the new testhelper directory.
Here's the diff for vendor.json


```
fowler@snoopy:~/Code/src/github.com/terraform-providers/terraform-provider-openstack$ git diff vendor/vendor.json
diff --git a/vendor/vendor.json b/vendor/vendor.json
index 0a4b163..99fe6bd 100644
--- a/vendor/vendor.json
+++ b/vendor/vendor.json
@@ -342,6 +342,18 @@
                        "revision": "0f64da0e36de86a0ca1a8f2fc1b0570a0d3f7504",
                        "revisionTime": "2017-03-10T01:59:53Z"
                },
+               {
+                       "checksumSHA1": "lW+ldKF07Zz/C7WYPTchfr1USrQ=",
+                       "path": "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces",
+                       "revision": "282f25e4025de0a42015d2e2b5faef1d920aad3c",
+                       "revisionTime": "2018-05-15T01:47:05Z"
+               },
+               {
+                       "checksumSHA1": "tHW4ajx6KpesbhMfizNicJ6g9Ks=",
+                       "path": "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/attachinterfaces/testing",
+                       "revision": "282f25e4025de0a42015d2e2b5faef1d920aad3c",
+                       "revisionTime": "2018-05-15T01:47:05Z"
+               },
                {
                        "checksumSHA1": "y49Ur726Juznj85+23ZgqMvehgg=",
                        "path": "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/availabilityzones",
@@ -678,6 +690,18 @@
                        "revision": "0f64da0e36de86a0ca1a8f2fc1b0570a0d3f7504",
                        "revisionTime": "2017-03-10T01:59:53Z"
                },
+               {
+                       "checksumSHA1": "b8tetZflGv6Pg22zYyodQGce7nc=",
+                       "path": "github.com/gophercloud/gophercloud/testhelper",
+                       "revision": "282f25e4025de0a42015d2e2b5faef1d920aad3c",
+                       "revisionTime": "2018-05-15T01:47:05Z"
+               },
+               {
+                       "checksumSHA1": "XHWKJUGCvXQNDmRrev0ZOVcISSM=",
+                       "path": "github.com/gophercloud/gophercloud/testhelper/client",
+                       "revision": "282f25e4025de0a42015d2e2b5faef1d920aad3c",
+                       "revisionTime": "2018-05-15T01:47:05Z"
+               },
                {
                        "checksumSHA1": "ZguAwNlZc+b2wSb+P4WddyBhNBA=",
                        "path": "github.com/gophercloud/utils/openstack/clientconfig",
```

jtopjian commented:
I apologize for the trouble with this :(

I've tried reproducing this on my workstation and it looks to be working correctly. This might be a problem with the version of `govendor` you're using. 

Did you install `govendor` via

```
go get -u github.com/kardianos/govendor
```

or by downloading a release? If the latter, try running the above Go command instead.

Elethiomel commented:
OK, I've gone head and used govendor and committed the testhelper directory too. Minor conflict in vendor.json fixed.

Elethiomel commented:
Documentation uploaded. Now to get tests working

jtopjian commented:
@Elethiomel I wanted to touch base and see if you needed any help with this PR? By all means, take all the time you need - I just wanted to see if you needed any help :)

Elethiomel commented:
@jtopjian Hey Joe, I've been away at the Openstack Conference in Vancouver so I've not been able to touch this for a week or so. I have got some stuff for tests and setting fixed ips on the ports in the pipline though! I hope to push it in the next few days.

Elethiomel commented:
@jtopjian I might literally have been sitting beside you at one point. There was a lightening talk on Terraform I was at!
I've now committed the support for a fixed_ip on the attached interface. I'm now working on a test for that, but I need to know what the CIDR is on the OS_NETWORK_ID you guys use on the test boxes as the fixed_ip must be in that CIDR for the tests to pass.

Elethiomel commented:
I think we're ready for a real review now. The test for fixed_ip will probably fail on the openlab check due to the aforementioned issue with the fixed_ip needing to be in a range in the OS_NETWORK_ID network. That is of-course unless you have a 10.119/16 network or a 10/8 one set up :)

jtopjian commented:
@Elethiomel Yep, I was at the lightning talk. Then I had to rush to the _other_ Terraform talk (more focused on Octavia) happening at the same time. :)

wrt to `OS_NETWORK_ID`, it's best to consider the subnet unknown. But a workaround is to do something like the following: https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_compute_instance_v2_test.go#L914-L944

Here you can see an instance being created with two NICs: one on the `OS_NETWORK_ID` and one on a custom defined network. On that network, you can have full control over the IP used.

I don't think it's a requirement to have the instance always on the default network. If all of the existing tests do this, it might be of "just in case" than a technical requirement.

Elethiomel commented:
@jtopjian Silly me, I should have thought of that! OK, it's now "working on my machine" (TM). Let's hope the openlab tests pass now

Elethiomel commented:
@jtopjian  Hrrrm, not sure why that failed. The logs are very very verbose. Any ideas?

jtopjian commented:
@Elethiomel Yeah, the xunit failures aren't really readable. The `job-output.txt.gz` used to have cleanly formatted plain txt to read, but that output is no longer showing (side note to anyone else reading this: I've reported the issue to some of the OpenLab team).

The JSON output has somewhat readable formatting. From there, I can see:

```
openstack_compute_attach_interface_v2.ai_1: Invalid request due to incorrect syntax or missing required parameters.
```

I've just merged a Gophercloud patch that will make this error message more helpful by including the actual error. I'll need to do a vendor update to Terraform, though. I can get to this in the next few days.

Until then, using `TF_LOG=DEBUG OS_DEBUG=1` when invoking Terraform will show the actual JSON calls which will include the full error message. Of course, if this is working on your side, that won't help. I'll run this on the clouds I have access to and see if I can reproduce the error.

jtopjian commented:
recheck

jtopjian commented:
@Elethiomel I dug into the logs a little more and the error doesn't seem right:

```
"2018/05/31 12:05:39 [DEBUG] Creating interface attachment: attachinterfaces.CreateOpts{PortID:\"\", NetworkID:\"1b4fb26f-2cbe-42e3-b2f7-655834d66ab5\", FixedIPs:[]attachinterfaces.FixedIP{attachinterfaces.FixedIP{SubnetID:\"\", IPAddre
ss:\"10.119.100.100\"}}}"
```

It's not using the right `FixedIP`. This might be an issue with OpenLab somehow. I'm re-running the test so we'll see what happens.

jtopjian commented:
Well this time around the right IP was used, but still failing:

```
2018/06/03 06:10:46 [DEBUG] Creating interface attachment: attachinterfaces.CreateOpts{PortID:\"\", NetworkID:\"6c14525c-f603-4bca-983e-858a48044ca0\", FixedIPs:[]attachinte
rfaces.FixedIP{attachinterfaces.FixedIP{SubnetID:\"\", IPAddress:\"192.168.1.100\"}}}
```

I'm going to get that Gophercloud patch vendored in and then we might better be able to determine the cause of failure.

jtopjian commented:
@Elethiomel I just merged #336 which should provide a more detailed error message. Would you be able to rebase your branch with upstream master? Let me know if you need help with the process.

Elethiomel commented on openstack/resource_openstack_compute_attach_interface_v2.go:
```diff
@@ -0,0 +1,239 @@
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
```
Fixed in pending commit

Elethiomel commented on openstack/resource_openstack_compute_attach_interface_v2.go:
```diff
@@ -0,0 +1,239 @@
+		return fmt.Errorf("Must set one of network_id and port_id")
+	}
+
+	if networkId != "" && portId != "" {
```
Fixed in pending commit

Elethiomel commented:
@jtopjian I'm still alive :) I've been really busy with other things, but I'm going to try get this finished off soon. I'll be pushing some changes soon. Not everything is tested or finished yet though

Elethiomel commented on website/docs/r/compute_attach_interface_v2.html.markdown:
````diff
@@ -0,0 +1,182 @@
+
+```
+
+### Basic attachment of a network to to a single instance, auto-creating a port, and specifiying an IP address. NOTE: You must specify a network_id when using fixed_ip and that ip must lie in the range available on the network
````
Fixed in upcoming commit

Elethiomel commented on website/docs/r/compute_attach_interface_v2.html.markdown:
````diff
@@ -0,0 +1,182 @@
+```
+
+
+### Basic attachment of a single interface (a port) to a single instance
````
Fixed in upcoming commit

Elethiomel commented on website/docs/r/compute_attach_interface_v2.html.markdown:
````diff
@@ -0,0 +1,182 @@
+
+```
+
+### Attaching multiple ports to a single instance
````
Fixed in upcoming commit

jtopjian commented:
@Elethiomel Finally see the error!

```
Bad request with: [POST http://192.168.0.5/compute/v2.1/servers/69e4d2fc-dacf-40d0-b0bb-fa99d7b3dd75/os-interface], error message: {"badRequest": {"message": "Invalid input for field/attribute 0. Value: {u'subnet_id': u'', u'ip_address': u'192.168.1.100'}. Additional properties are not allowed (u'subnet_id' was unexpected)", "code": 400}}
```

I have a fix for this here:

https://github.com/gophercloud/gophercloud/issues/1134
https://github.com/gophercloud/gophercloud/pull/1135

Once it's merged, one final vendor update should do the trick.

Elethiomel commented:
I wonder why it works on my OpenStack Cluster without these patches.

What do you want to do about the instance_id etc? I could store them in the attachment ID. What do you think?

jtopjian commented on openstack/resource_openstack_compute_attach_interface_v2.go:
```diff
@@ -0,0 +1,239 @@
+	log.Printf("[DEBUG] Retrieved interface attachment: %#v", attachment)
+
+	d.Set("port_id", attachment.PortID)
+	d.Set("network_id", attachment.NetID)
```
My mistake. Since `instance_id` is a required argument of this resource, we don't need to set it.

For `region`, look at the other resources for how `d.Set("region", ...)` is being done.

And for `fixed_ip`, let's make sure `attachment.FixedIPs` has a length of 1 before setting so we don't accidentally hit an out of bounds error.

jtopjian commented:
> I wonder why it works on my OpenStack Cluster without these patches.

It works for me, too, but I was finally able to reproduce the issue on a Queens test cloud I have. #1135 has been merged, so that's good to go :)

> What do you want to do about the instance_id etc?

I left an inline comment about this. My mistake about `instance_id` - sorry for confusing things.

jtopjian commented on openstack/resource_openstack_compute_attach_interface_v2.go:
```diff
@@ -0,0 +1,251 @@
+		d.Set("fixed_ip", attachment.FixedIPs[0])
+	} else {
+		log.Printf("[DEBUG] Length of attachment.FixedIPs is %d, not 1 - skipping Set of fixed_ip field", len(attachment.FixedIPs))
+	}
```
This is my mistake again. Let's just remove this entirely and not set the `fixed_ip` at all here.

There are two issues that can come out of this:

1. A user attaches a port with multiple fixed IPs
2. The `FixedIP` result also contains a `SubnetID` which will also be dropped

There are ways around both of these, but it's a bit complicated and involves adding a new attribute to the schema. For a first commit of this resource, we can safely omit it and add it later.

So the only missing `d.Set` was `region` - again, my apologies.

jtopjian commented on website/docs/r/compute_attach_interface_v2.html.markdown:
````diff
@@ -0,0 +1,182 @@
+```
+
+
+## Argument Reference
````
Need to add `region` to Argument Reference. Just copy and paste it from another file - it has a standard description.

mcanevet commented:
@Elethiomel any new regarding this PR? This is definitely a feature I'd like

jtopjian commented:
I think this is ready to go - it just needs rebased and some vendor cleanup. I can take over.

jtopjian commented:
This was merged as #470 

@Elethiomel Thank you so much for your work on this!
